### DISCUSSION TITLE
Advanced Regression Technique

### DISCUSSION THREAD

**Discussion by mirfani28** [teamalgoritma/community#459] [Q&A] [answered]
Hello Guys.
Allow me to introduce myself, I'm Irfani, a student from Theia Batch in Algoritma Course.

I stumble across this interesting article : https://www.kaggle.com/code/serigne/stacked-regressions-top-4-on-leaderboard
In that article, the writer uses several regression techniques such as LASSO Regression, Elastic Net Regression, XGBoost, LightGBM, and many more to make models.

After that, he stacked the models and then ensemble a few more models with the stacked model he made before.

- I understand the basic concept of linear regression but these terms are strange for me. Why are there so many linear regression techniques? or some of those regressions are not linear? Is there also exponential regression in that model?

- How does the ensemble method work? Why is it not exclusive for classification problems? I thought it was exclusive only to classification models because I only heard that term in Classification Class in the last two weeks (CMIIW )

- One last question, can we do advanced regression techniques in R? What are the equivalent libraries for all the regression techniques mentioned in the article above.

Thank You
Best Regards

**Comment by VictorNugraha** [comment]
Dear @mirfani28,

Terima kasih atas pertanyaan yang disampaikan, pertanyaan Anda sudah kami terima dan sedang kami akan mencoba untuk membaca article yang dilampirkan terlebih dahulu dan secepatnya akan kami berikan jawaban untuk beberapa pertanyaan yang diberikan.

**Comment by onlyphantom** [comment, marked as answer]
HI Irfani,

Thanks for the putting down your questions and hopefully I can help with providing some perspective on your queries above. 

1. Every model deals with a bias-variance trade-off. A model that has high bias, has low variance; and a complex model that has high variance is said to have low bias. Finding the balancing point is key to a good model. 

2. Linear regression using ordinary least squares has a "high bias, low variance" (technical term here, doesn't mean it's biased in the day-to-day sense of that word. High bias model can result in models that might overfit, since you have two coefficients (intercept and slope) that you're using to predict new values. If one of the two coefficient react strongly to an outlier value (recall the chapter on leverage and influence), that bias can have a significant effect on its overall prediction accuracy.

Compare that to a model like neural network for example that produces a grid of, say, 50000 coefficients. That is a high variance, low bias model.

3. You mentioned other regression methods like LASSO regression, ridge regression etc, and you asked if they are not linear / exponential. So Lasso regression and ridge regression are just **tiny modifications** of linear regression that aims to minimize the sum or  squared residuals, just as you have learned in the course materials. Lasso and Ridge (L1/L2) adds an extra term so instead of minimizing sum of squared residuals, you minimize sum of squared residuals plus the value of the slope. They are very similar in that it penalizes the model and force the values to shrink towards 0 just a bit more, in order to affect the bias-variance tradeoff mentioned in the points above. Why? To make it less prone to overfitting. 

These small modifications doesn't make a regression "not linear", it just make it less prone to overfit to the training data. Many of what you consider advanced regression follow this line of thought. In fact, all of machine learning surrounds this idea of managing a model's bias-variance tradeoff, so that it learns enough, but not try to learn so much it ends up memorizing instead of learning (high train accuracy, poor test accuracy). 

4. Ensemble is a very general technique. The idea is to compensate by getting a few learners and then through a mechanism (i.e such as voting), arrives at a better model.

Supposed I ask A, B, C, D, E, five students to predict if a loan is going to default or not. Supposed each one of them has only a 0.7 accuracy (7 out of 10 times, they will be correct). If I always side with the majority, I would be correct 8.34 out of 10 times (0.83 accuracy). This is because I  need 3 out of A, B, C, D, E to be correct, or 4 out of A, B, C, D, E to be correct, or 5 of them to be correct. I can use R to compute this probability:

```r
0.7^5 + (choose(5,4) * 0.7^4 * 0.3^1) + (choose(5,3) * 0.7^3 * 0.3^2)
[1] 0.83692
```

So broadly speaking, any techniques that combines multiple models to get an improved one is an ensemble method. In the example above, individually each student only has a 0.7 accuracy, but taking the majority I would yield an accuracy of 0.83. You don't need to use the majority either, supposed we ask the 5 to predict rainfall, and then I take the average of the 5 students' prediction, now I'm using ensemble to solve a regression problem instead of a classification. 

5. I'll go so far as to say you're more likely to find "advanced" linear regression packages in R than to find it in Python :) It's one of those areas where R wins comfortably. I would start by searching in CRAN for example using the terms you mentioned [like xgboost](https://search.r-project.org/?P=xgboost&SORT=&HITSPERPAGE=10&DB=cran-info&DEFAULTOP=and&FMT=query&xP=ZBxgboost%09ZFxgboost%09ZSxgboost%09Zxgboost&xDB=cran-help%2Fcran-readme&xFILTERS=.%7E%7E4294967295) and [lasso](https://search.r-project.org/?P=lasso&SORT=&HITSPERPAGE=10&DB=cran-info&DEFAULTOP=and&FMT=query&xP=ZBxgboost%09ZFxgboost%09ZSxgboost%09Zxgboost&xDB=cran-info&xFILTERS=.%7E%7E4294967295) there are 233 packages with a match. R users are very spoiled for choices when it comes to advanced statistical techniques and libraries compared to Python users.

**Reply by mirfani28** [reply]
Hi Mr Samuel Chan, I really appreciate your thorough answer. It makes a lot of sense now
Another questions about this :

- I'm interested in the variance-bias tradeoff that you mentioned before. Let's say we make a model from a sample with 10 independent predictors compared with a model with 5 independent predictors, does the model with 10 independent predictors always produce a better model? Assume that the number of samples are the same (but not identical) but both samples are taken at the same time. 

- How data collecting is done and how we decide how many variables we will take? Do we have to make base assumptions first about the model that we will make or are we just collecting all the things that we "think" maybe significant to the target variable? (Assuming that we have tons of money and don't worry about the man-hour for data collecting)

- In what real-world examples can we use exponential regression? and when we decide to use it? Do we have to use linear regression first and after certain tests or evaluations we can start to think about exponential regression? 

- Can we use exponential regression for example to predict traffic? Like in my experience when I was living in a big city, the number of vehicles on the street is really really different between 15 minutes during the start of rush-hour. Like 15 minutes late in departure can result in 2 hours late in arrival. 

Thanks

**Reply by mirfani28** [reply]
Ups, sorry I forgot, another question.

- Can you explain a little how log transformation is used to process skewed distribution of samples?

**Reply by dwiknrd** [reply]
Halo [mirfani28](https://github.com/mirfani28),

Thank you for asking, I will give some insight from what you asked above, hopefully it will be useful

1. According to research, if the model is tuned with too much to the data, then it could essentially memorise the data, and that causes model overfitting, which causes high error rates for unseen data. If we are overfitting, we get wrong predictions and lose the focus on what’s actually important. An overfitting model implies that you have low bias and high variance and more data is not going to solve your problem. By placing too much emphasis on each data point, data scientists have to deal with a lot of noise and, therefore, lose sight of what’s really important. So adding more data points to the training set will not improve the model performance. The more number of independent variables does not mean our model will be better. Not necessarily the variables used are significant to the model. So it is necessary to do several ways such as feature selection.
2. It really depends on the problem. More is always better. But the most common way to define whether a data set is sufficient is to apply a 10 times rule. This rule means that the amount of input data (i.e., the number of examples) should be ten times more than the number of degrees of freedom a model has. Usually, degrees of freedom mean parameters in your data set. So, for example, if your algorithm distinguishes images of cats from images of dogs based on 1,000 parameters, you need 10,000 pictures to train the model.
3. Exponential regression is used to model situations in which growth begins slowly and then accelerates rapidly without bound, or where decay begins rapidly and then slows down to get closer and closer to zero.

I hope this helps :)

**Reply by mirfani28** [reply]
Thanks a lot Mbak @dwiknrd for your detailed answer :)

Any recommended article for exponential regression in data modelling?
